### ISSUE TITLE
Doesn't recognize Google Photos compressed version as a duplicate

### ISSUE THREAD

**Issue by mtlynch** [photoprism/photoprism#1182]
I have a collection of photos on my local drive and a collection on Google Photos. They largely overlap, but there are some photos unique to each. I'd like to use Photoprism to merge them back together and de-dupe the overlap, but it seems to not recognize that they're duplicates.

Here's a typical example. The original is the version straight off the camera, and the Google Photos version has been uploaded to Google and downloaded via Google Takeout.

| Version | Size | Hash |
|----------|------|------|
| [Original](https://storage.googleapis.com/public-scratch-s99r16fnoil244tewk85/photoprism/original/DSC01177.JPG) | 1.7 MB | `50af9ce2a535c817609ace71d8a9d73119b53297` |
| [Google Photos](https://storage.googleapis.com/public-scratch-s99r16fnoil244tewk85/photoprism/google-photos/DSC01177.JPG) | 396.1 KB | `c3bdddce278b972f7a752e31359e69003a384511` |

From looking at the files, it seems like Google Photos compressed the photo even though it looks identical to my eye, but this changed the hash.

Is it possible to adjust how Photoprism evaluates duplicates so that it catches cases like these?

**Comment by Sblop** [comment]
Is it recognized as a version of the same and put in the same stack?

**Comment by mtlynch** [comment]
@Sblop I'm on 210222-ac5a9d5e-Linux-x86_64. I don't see a Stacks option in settings or in the UI.

**Comment by graciousgrey** [comment]
You can enable stacking by time & place, if location and time match in both versions of your files they will be stacked.
See: https://docs.photoprism.org/user-guide/organize/stacks/ and https://docs.photoprism.org/user-guide/settings/library/#stacks.

**Comment by mtlynch** [comment]
@graciousgrey Thanks for the clarification!

I do see stacks, but I'm not quite sure how to use them. It shows that photos are stacked, but when I click the stack, it only shows a single photo:

https://user-images.githubusercontent.com/7783288/114481693-12c16100-9bd3-11eb-8cd9-b7545b67b0f7.mp4

Is that a bug or am I not understanding how to use stacks?

In any case, most of the duplicate photos I see don't show up in stacks. Is there a way to hide duplicates like the example I shared where the duplicate is a compressed version of the original?

**Comment by graciousgrey** [comment]
@mtlynch if you click the sequence symbol (upper left corner) on the photo you can see the photos belonging to this stack by using the arrows (https://docs.photoprism.org/user-guide/organize/stacks/#browse-through-related-files). 

PhotoPrism does duplicate detection using hashes, so this won't work here because the hashes are different. 
Do the compressed images from Google have any metadata that show a relation to the original? Like the same date, a similar filename?

**Comment by mtlynch** [comment]
>if you click the sequence symbol (upper left corner) on the photo you can see the photos belonging to this stack by using the arrows (https://docs.photoprism.org/user-guide/organize/stacks/#browse-through-related-files).

@graciousgrey Can you take a look at [the video I posted upthread](https://github.com/photoprism/photoprism/issues/1182#issuecomment-818354190)? I'm clicking the sequence symbol in the stacks view, but it's only showing one image.

>Do the compressed images from Google have any metadata that show a relation to the original? Like the same date, a similar filename?

Yes, the image files are posted [upthread](https://github.com/photoprism/photoprism/issues/1182#issue-854067746). They have identical filenames and identical timestamps in the metadata. These two photos do not appear in the stacks view.

**Comment by graciousgrey** [comment]
> Can you take a look at the video I posted upthread? I'm clicking the sequence symbol in the stacks view, but it's only showing one image.

I just realized that it is a video. Seems like the two files stacked have no visible difference. I guess one is your original and the other one is the compressed one from google. If you go to the files tab (https://docs.photoprism.org/user-guide/organize/edit/#edit-files) of this photo you can check the names and sizes of the files that are stacked. 

> They have identical filenames and identical timestamps in the metadata. These two photos do not appear in the stacks view.

We do not stack files by identical filenames, as there is a high risk that many files that are not related are stacked e.g. `photo.jpg` or `IMG_1234.jpg`.
We do stack by time AND place. 
So what happens is stacking original and compressed versions works only for photos that have time and place information like the ones from the video, as far as I can see all of those have a location.

Photos without a location like the ones you mentioned above won't be stacked automatically.

What you could do as a workaround is searching for photos with the country filter set to Unknown and sorting the results by similarity.

**Comment by mtlynch** [comment]
>I just realized that it is a video. Seems like the two files stacked have no visible difference. I guess one is your original and the other one is the compressed one from google. If you go to the files tab (https://docs.photoprism.org/user-guide/organize/edit/#edit-files) of this photo you can check the names and sizes of the files that are stacked.

It seems like it's just a single file in the stack for some reason. I looked at the file information, and it only shows one file (see video below):

https://user-images.githubusercontent.com/7783288/114641111-7b760f80-9c9f-11eb-9ceb-956fd21be6a7.mp4

Should I file a separate bug?

>What you could do as a workaround is searching for photos with the country filter set to Unknown and sorting the results by similarity.

I'm not sure this would be practical in my case. I have thousands of duplicates between the Google Photos version and the originals. Would I need to go through each duplicate by hand and mark the Google Photos version as redundant?

**Comment by dror3go** [comment]
@mtlynch When in the editing page under the Files tab - notice that it may contain multiple photos (hence "File**s**"). You'll need to scroll down, or click on the header with the file's name (below the Details/Labels/Files/settings tabs and above the Preview) to toggle the file's info and see the rest of the photos.

**Comment by lastzero** [comment]
Yes, these are two files that looks pretty much the same... scroll down in the files tab to see the other file.

**Comment by mtlynch** [comment]
Thanks, @lastzero and @dror3go. It does show other files in the stack in that tab.

Coming back to the original issue, is there a good way to use Photoprism to solve the problem of thousands of duplicate photos between a Google Takeout export and the originals? I understand that I can see certain photos in stacks and that I can sort by similarity, but is there a way to resolve duplicates in bulk?

The photos in these cases should have a near identical content hash, and they have identical filenames and camera metadata, though most of them lack location information.

**Comment by graciousgrey** [comment]
@mlynch Unfortunately at the moment duplicates are only recognized by hash.

I see the following to dos:

1) Stack compressed and original images e.g. when name and date match, size differs AND there is a very high similarity. This way the photo appears only once in the photo view and the user can then decide what to do with the stacked files. Keep both or delete one.

2) In a next step we could think about an optional setting to automatically delete the similar image with lower or higher quality. But we need to be careful with this as we don't want to delete images that are not a compressed but e.g. a edited version of the original.

What do you think?

If you need a solution now and you are familiar with databases you could search for the files with the same name and then set a deletedAt date for the photo that belongs to the file with the size you don't want to keep. Photos with a deletedAt date appear in archive and can be deleted from there via the UI.

**Comment by mtlynch** [comment]
@graciousgrey that sounds good to me. Should I file separate bugs? Alternatively, feel free to edit my original report.

My only note is that I'd personally prefer to just *hide* duplicates rather than delete them. I regularly dump data from Google Takeout, so if I delete a duplicate, it will just be added back on the next Takeout export. If I hide them, they'll stay hidden next time I dump from Google. That may be an unusual workflow, so I understand if Photoprism's solution depends on deleting duplicates.

**Comment by graciousgrey** [comment]
@mlynch thank you for your feedback! I will just edit this issue  :)
